File: base_demo/scripts/service_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import rclpy
from rclpy.node import Node
from base_demo.srv import SetTargetDetec


class MinimalClientAsync(Node):

    def __init__(self):
        super().__init__('service_client')
        self.cli = self.create_client(SetTargetDetec, 'target_detection')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('service not available, waiting again...')
        self.req = SetTargetDetec.Request()

    def send_request(self):
        self.req.name = "box"
        self.future = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.future)
        return self.future.result()


def main():
    rclpy.init()
    minimal_client = MinimalClientAsync()
    response = minimal_client.send_request()
    if response.success:
        minimal_client.get_logger().info('Target detection succeeded!')
        minimal_client.get_logger().info('Pose: x = %f ,y= %f, z= %f' 
        %(response.pose.position.x,response.pose.position.y,response.pose.position.z))
    else:
        minimal_client.get_logger().info('Can not find the target!')

    minimal_client.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: base_demo/src/service_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "base_demo/srv/set_target_detec.hpp"
#include <chrono>
#include <cstdlib>
#include <memory>

using namespace std::chrono_literals;

int main(int argc, char **argv){
  // 初始化ROS节点
  rclcpp::init(argc, argv);
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "service_client node is Ready!");
  
  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("service_client");

  rclcpp::Client<base_demo::srv::SetTargetDetec>::SharedPtr client =
    node->create_client<base_demo::srv::SetTargetDetec>("target_detection");
  auto request = std::make_shared<base_demo::srv::SetTargetDetec::Request>();
  request->name = "box";

  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "service not available, waiting again...");
  }

  auto result = client->async_send_request(request);
  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    auto res = result.get();
    if(res->success){
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Target detection succeeded!");
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Robot pose x : %.2fm; y : %.2fm; z : %.2fm", res->pose.position.x, res->pose.position.y, res->pose.position.z);
    }else{
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Can not find the target!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Failed to call service target_detection");
  }

  rclcpp::shutdown();
  return 0;
}
